src/board.rs

```diff
@@ -39,83 +41,89 @@ impl Board {
         }
     }
 
-    fn get_back_rank(player: Player) -> [Piece; 8] {
+    fn get_back_rank(player: Player) -> [Option<Piece>; 8] {
         [
-            Piece::Rook(player),
-            Piece::Knight(player),
-            Piece::Bishop(player),
-            Piece::Queen(player),
-            Piece::King(player),
-            Piece::Bishop(player),
-            Piece::Knight(player),
-            Piece::Rook(player),
+            Some(Piece::Rook(player)),
+            Some(Piece::Knight(player)),
+            Some(Piece::Bishop(player)),
+            Some(Piece::Queen(player)),
+            Some(Piece::King(player)),
+            Some(Piece::Bishop(player)),
+            Some(Piece::Knight(player)),
+            Some(Piece::Rook(player)),
         ]
     }
 
-    pub fn is_in_bounds(position: Position) -> Result<(), ChessError> {
+    pub fn is_in_bounds(position: Position) -> ChessResult<()> {
         if position.x > 7 || position.y > 7 {
             Err(ChessError::OutOfBounds)
         } else {
             Ok(())
         }
     }
 
-    fn is_valid_move(&self, from: Position, to: Position) -> Result<(), ChessError> {
-        Self::is_in_bounds(from)?;
-        Self::is_in_bounds(to)?;
-
-        let piece = &self.squares[from.x][from.y];
-
-        match *piece {
-            Piece::Pawn(..) => {}
-            Piece::Knight(..) => {}
-            Piece::Bishop(..) => {}
-            Piece::Rook(..) => {}
-            Piece::Queen(..) => {}
-            Piece::King(..) => {}
-            Piece::None => return Err(ChessError::InvalidMove),
-        };
-        Ok(())
+    fn is_piece_some(&self, position: Position) -> ChessResult<()> {
+        if self.get_piece(position)?.is_none() {
+            Err(ChessError::NoPieceAtPosition)
+        } else {
+            Ok(())
+        }
     }
 
-    pub fn move_piece(&mut self, from: Position, to: Position) -> Result<(), ChessError> {
-        // Check if 'from' and 'to' positions are in bounds
-        Self::is_in_bounds(from)?;
+    fn is_move_valid(&self, from: Position, to: Position) -> ChessResult<()> {
+        //we need ending position to be inbound
         Self::is_in_bounds(to)?;
 
-        if self._get_piece(from).is_none() {
-            return Err(ChessError::NoPieceAtPosition);
+        //moving piece must be in bound and not `None`
+        self.is_piece_some(from)?;
+
+        if let Some(piece) = self._get_piece(from) {
```

Comment:
Do we need this whole `if .. else` block? Isn't it already checked by `is_piece_some(from)`?

Reply:
Oh wait I guess this is a place holder for actually checking the moves? Ultimately I want to have them pre-computed in `self.moves` and you just check if `(from, to)` is in the set.